[WIP] Streaming encoder and CTC

pzelasko opened the pull request:
Highlights:
- `StreamingE2E` class with an API for streaming computation
- `Encoder`s `forward` method extended to be able to propagate state between audio chunks
- no streaming attention decoder yet

sw005320 commented on espnet/nets/pytorch_backend/encoders.py:
```diff
@@ -100,25 +105,37 @@ def __init__(self, idim, elayers, cdim, hdim, dropout, typ="blstm"):
+        return xs_pad, ilens, states  # x: utt list of frame x dim
+
+
+def zero_backward_rnn_state(states):
```
It's better to comment about the role of this function.

pzelasko replied:
:+1:

sw005320 commented:
@kan-bayashi, could you review the PR?

kan-bayashi commented:
I will do it this midnight or tomorrow.

kan-bayashi commented on espnet/nets/pytorch_backend/e2e_asr.py:
```diff
@@ -375,9 +371,103 @@ def calculate_all_attentions(self, xs_pad, ilens, ys_pad):
+            return torch.cat(self._encoder_states, dim=0), torch.cat(self._ctc_posteriors, dim=0)
+
+        def select_unprocessed_windows(window_tensors):
+            last_offset = self._offset
```
Is it necessary to update `self._offset?` at the end of this function?

kan-bayashi replied:
Should `self._encoder_states` and `self._ctc_posteriors` used for decoding be removed?
Or why don’t you prepare `self.reset()` to clear the stored them?

pzelasko replied:
This is the next step - I want to submit (at least a prototype of) streaming attention decoder in a follow-up PR. Then I'll add the necessary operations on the offset, or just clean up the "old" states and posteriors as suggested.

kan-bayashi commented on espnet/nets/pytorch_backend/e2e_asr.py:
```diff
@@ -375,9 +371,103 @@ def calculate_all_attentions(self, xs_pad, ilens, ys_pad):
+        """
+        h, lpz = self._input_window_for_decoder(use_all=True)
+
+        self._last_recognition = self._e2e.dec.recognize_beam(h, lpz, self._recog_args, self._char_list, self._rnnlm)
```
Is there any reason why this method does not return recognition results directly?

pzelasko replied:
Good question :) I anticipate the this kind of interface will make the implementation of streaming attention decoder useful, but for this PR it's not needed at all - I'll remove these pieces and if I need them, I'll submit them again in the next PR.

pzelasko commented:
I've addressed the code review suggestions. I hope to put out another PR this week which implements a prototype of the streaming attention decoder and then I'll introduce the offset/past state management in `StreamingE2E`.

sw005320 commented:
Thanks @pzelasko!
